File: tests/samples/java/script_4.java
// Test script 4

import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.context.Scope;

import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Collectors;
import java.util.HashMap;

public class Script4 {
    private static final Tracer tracer = OpenTelemetry.getGlobalTracer("script_4_tracer");

    /**
     * Simulate async data fetch with delay.
     */
    private CompletableFuture<Map<String, List<Integer>>> fetchMockData() {
        return CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(100); // Simulate network delay
                Map<String, List<Integer>> result = new HashMap<>();
                result.put("data", List.of(1, 2, 3, 4, 5));
                return result;
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        });
    }

    /**
     * Simulate async data processing.
     */
    private CompletableFuture<List<Integer>> processMockData(Map<String, List<Integer>> data) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                Thread.sleep(50); // Simulate processing time
                return data.get("data").stream()
                    .map(x -> x * 2)
                    .collect(Collectors.toList());
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new RuntimeException(e);
            }
        });
    }

    /**
     * Async function that fetches data with telemetry.
     */
    public CompletableFuture<List<Integer>> asyncFetchData() {
        Span span = tracer.spanBuilder("fetch_data").startSpan();
        
        return CompletableFuture.supplyAsync(() -> {
            try (Scope scope = span.makeCurrent()) {
                span.setAttribute("request_type", "async");
                
                return fetchMockData()
                    .thenCompose(rawData -> {
                        span.setAttribute("data_size", rawData.get("data").size());
                        
                        return processMockData(rawData)
                            .thenApply(processedData -> {
                                span.addEvent("data_processed", Attributes.builder()
                                    .put("input_size", rawData.get("data").size())
                                    .put("output_size", processedData.size())
                                    .build());
                                return processedData;
                            });
                    })
                    .exceptionally(throwable -> {
                        span.recordException(throwable);
                        throw new CompletionException(throwable);
                    })
                    .join();
            } finally {
                span.end();
            }
        });
    }

    // Example usage
    public static void main(String[] args) {
        Script4 script = new Script4();
        script.asyncFetchData()
            .thenAccept(result -> System.out.println("Processed data: " + result))
            .exceptionally(throwable -> {
                System.err.println("Error: " + throwable.getMessage());
                return null;
            });
    }
}

File: tests/samples/python/script_4.py
# Test script 4

import asyncio

from opentelemetry import trace

from typing import List
from typing import Dict

tracer = trace.get_tracer("script_4_tracer")


async def fetch_mock_data() -> Dict[str, List[int]]:
    """Simulate async data fetch with delay."""
    await asyncio.sleep(0.1)  # Simulate network delay
    return {"data": [1, 2, 3, 4, 5]}


async def process_mock_data(data: Dict) -> List[int]:
    """Simulate async data processing."""
    await asyncio.sleep(0.05)  # Simulate processing time
    return [x * 2 for x in data["data"]]


async def async_fetch_data():
    """
    Async function that fetches data with telemetry.
    """
    async with tracer.start_as_current_span("fetch_data") as span:
        span.set_attribute("request_type", "async")
        try:
            raw_data = await fetch_mock_data()
            span.set_attribute("data_size", len(raw_data["data"]))

            processed_data = await process_mock_data(raw_data)
            span.add_event(
                "data_processed",
                {
                    "input_size": len(raw_data["data"]),
                    "output_size": len(processed_data),
                },
            )

            return processed_data
        except Exception as e:
            span.record_exception(e)
            raise
